hoss/dimension_utilities.py

```diff
@@ -55,54 +65,329 @@ def is_index_subset(message: Message) -> bool:
 
     # Create bounds variables if necessary.
-    add_bounds_variables(required_dimensions_nc4, required_dimensions, varinfo, logger)
+    add_bounds_variables(required_variables_nc4, required_variables, varinfo, logger)
+    return required_variables_nc4
+
+
+def get_override_projected_dimension_name(
+    varinfo: VarInfoFromDmr,
+    variable_name: str,
+) -> str | None:
+    """returns the x-y projection variable names that would
+    match the geo coordinate names. The `latitude` coordinate
+    variable name gets converted to 'projected_y' dimension scale
+    and the `longitude` coordinate variable name gets converted to
+    'projected_x'
+
+    """
+    override_variable = varinfo.get_variable(variable_name)
+    projected_dimension_name = None
+    if override_variable is not None:
+        if override_variable.is_latitude():
+            projected_dimension_name = 'projected_y'
+        elif override_variable.is_longitude():
+            projected_dimension_name = 'projected_x'
+    return projected_dimension_name
+
+
+def get_override_projected_dimensions(
+    varinfo: VarInfoFromDmr,
+    variable_name: str,
+) -> list[str]:
+    """
+    Returns the projected dimensions names from coordinate variables
+    """
+    coordinate_variables = get_coordinate_variables(varinfo, [variable_name])
+    if coordinate_variables:
+        override_dimensions = []
+        for coordinate in coordinate_variables:
+            override_dimensions.append(
+                get_override_projected_dimension_name(varinfo, coordinate)
+            )
+    else:
+        # if the override is the variable
+        override = get_override_projected_dimension_name(varinfo, variable_name)
+        override_dimensions = ['projected_y', 'projected_x']
+        if (override is None) or (override not in override_dimensions):
+            override_dimensions = []
+    return override_dimensions
+
+
+def get_coordinate_variables(
+    varinfo: VarInfoFromDmr,
+    requested_variables: Set[str],
+) -> list[str]:
+    """This method returns coordinate variables that are referenced
+    in the variables requested. It returns it in a specific order
+    [latitude, longitude]
+    """
+
+    coordinate_variables_set = varinfo.get_references_for_attribute(
+        requested_variables, 'coordinates'
+    )
+    coordinate_variables = []
+    contains_latitude = False
+    contains_longitude = False
+    for coordinate in coordinate_variables_set:
+        if varinfo.get_variable(coordinate).is_latitude():
+            coordinate_variables.insert(0, coordinate)
+            contains_latitude = True
+        elif varinfo.get_variable(coordinate).is_longitude():
+            coordinate_variables.insert(1, coordinate)
+            contains_longitude = True
+
+    if coordinate_variables:
+        if not contains_latitude:
+            raise MissingCoordinateDataset('latitude')
+        if not contains_longitude:
+            raise MissingCoordinateDataset('longitude')
```

Comment:
Another thought (probably more important) - what about multiple grids? SPL3FP has global and polar grids, so has 2 latitude and 2 longitude coordinate variables. Your current implementation would get you something that is a 4-element list, and that ordering would depend entirely on the random ordering of the output of `VarInfoFromDmr.get_references_for_attribute`. That seems troubling. A big question here is how to make sure that the right latitudes and longitudes are being paired together.

When you write the unit tests for a lot of these functions, I definitely want to see what happens for a use-case when variables from different grids are being requested. We're going to need those tests to make sure things don't break.

Reply:
Good point. will add the multiple grid test cases

Reply:
> This function feels pretty convoluted now. I think it might be easier to do:
> 
> ```
> coordinate_variables_set = varinfo.get_references_for_attribute(
>     requested_variables, 'coordinates'
> )
> 
> latitude_coordinate_variables = [
>     coordinate
>     for coordinate in coordinate_variables_set
>     if varinfo.get_variable(coordinate).is_latitude()
> ]
> 
> longitude_coordinate_variables = [
>     coordinate
>     for coordinate in coordinate_variables_set
>     if varinfo.get_variable(coordinate).is_longitude()
> ]
> 
> if not latitude_coordinate_variables:
>     raise MissingCoordinateDataset('latitude')
> 
> if not longitude_coordinate_variables:
>     raise MissingCoordinateDataset('longitude')
> 
> return latitude_coordinate_variables, longitude_coordinate_variables
> ```
> 
> This reduces the number of local variables in the function, reduces the cyclomatic complexity and uses the more Python-native model of list-comprehensions. (Minor downside is two list comprehensions looping through the coordinates, but they should be a handful of elements, so I don't think we need to worry about performance)

updated the method - https://github.com/nasa/harmony-opendap-subsetter/pull/15/commits/802fe0ec04b853a7a7abb4718e2a05b0e8e6ed3c

Reply:
https://github.com/nasa/harmony-opendap-subsetter/pull/15#discussion_r1785092153

- not sure what that is referring to in the earlier comment

Reply:
> One other thought, though, what if the metadata attributes for the collection point to non existent variables? That will raise an exception that you currently aren't capturing.

if the coordinate variables don't exist, we will not do the override sections of the code

Reply:
Thanks for the update of the function.

> if the coordinate variables don't exist, we will not do the override sections of the code

I'm not sure that's quite true. The first time this function is called in the flow of the code is within `hoss.dimension_utilities::get_prefetch_variables`. The function gets called if there are no required dimensions. Then within this function you are looking at the `coordinates` metadata attribute on each required variable, and trying to retrieve information on each listed coordinate. If the metadata attribute is incorrect, then at that point, the code is asking `VarInfoFromDmr` to do something it can't do (`varinfo.get_variable(coordinate).is_longitude()` or `varinfo.get_variable(coordinate).is_latitude()`), and it will raise an exception. Specifically, it will say that `None` does not have a method `is_latitude` or `is_longitude`, because `varinfo.get_variable('something that doesn't exist')` will return `None`.

I think this failure case needs to be handled, most likely by catching the exception and re-raising a more user-friendly one.